# Capstone Project - Week 3
## Segmenting and Clustering Neighborhood in Toronto

### Part 1 -  Creating the initial data frame 

#### Install and import packages

In [1]:
!pip install bs4 # install beautifulsoup4

In [2]:
from bs4 import BeautifulSoup # this module helps in web scrapping.
import requests  # this module helps us to download a web page
import pandas as pd
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

#import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

#import requests # library to handle requests
#from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


#### Extract data from the page and load into a dataframe called df

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M" #the url containing the data
page  = requests.get(url).text # get the page
soup = BeautifulSoup(page,"html5lib") # create a soup object using the variable 'page'
table = soup.find('table') # the data is in the first table so we can use find method  

df = pd.DataFrame(columns=["PostalCode", "Borough", "Neighborhood"])
for row in table.findAll('td'):
    if row.span.text=='Not assigned':
        pass
    else:
        postalCode = row.p.text[:3]
        borough = (row.span.text).split('(')[0]
        neighbor = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        df = df.append({"PostalCode":postalCode, "Borough":borough, "Neighborhood":neighbor}, ignore_index=True)

df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

# On inspection, no neighbourhood has a 'not assigned valued' 



#### The resulting dataframe and its dimension (Part 1 result)

In [4]:
df

PostalCode                 Borough  \
0          M3A              North York   
1          M4A              North York   
2          M5A        Downtown Toronto   
3          M6A              North York   
4          M7A            Queen's Park   
5          M9A               Etobicoke   
6          M1B             Scarborough   
7          M3B              North York   
8          M4B               East York   
9          M5B        Downtown Toronto   
10         M6B              North York   
11         M9B               Etobicoke   
12         M1C             Scarborough   
13         M3C              North York   
14         M4C               East York   
15         M5C        Downtown Toronto   
16         M6C                    York   
17         M9C               Etobicoke   
18         M1E             Scarborough   
19         M4E            East Toronto   
20         M5E        Downtown Toronto   
21         M6E                    York   
22         M1G             Scarborough   
23         M4G               East York   
24         M5G        Downtown Toronto   
25         M6G        Downtown Toronto   
26         M1H             Scarborough   
27         M2H              North York   
28         M3H              North York   
29         M4H               East York   
30         M5H        Downtown Toronto   
31         M6H            West Toronto   
32         M1J             Scarborough   
33         M2J              North York   
34         M3J              North York   
35         M4J  East York/East Toronto   
36         M5J        Downtown Toronto   
37         M6J            West Toronto   
38         M1K             Scarborough   
39         M2K              North York   
40         M3K              North York   
41         M4K            East Toronto   
42         M5K        Downtown Toronto   
43         M6K            West Toronto   
44         M1L             Scarborough   
45         M2L              North York   
46         M3L              North York   
47         M4L            East Toronto   
48         M5L        Downtown Toronto   
49         M6L              North York   
50         M9L              North York   
51         M1M             Scarborough   
52         M2M              North York   
53         M3M              North York   
54         M4M            East Toronto   
55         M5M              North York   
56         M6M                    York   
57         M9M              North York   
58         M1N             Scarborough   
59         M2N              North York   
60         M3N              North York   
61         M4N         Central Toronto   
62         M5N         Central Toronto   
63         M6N                    York   
64         M9N                    York   
65         M1P             Scarborough   
66         M2P              North York   
67         M4P         Central Toronto   
68         M5P         Central Toronto   
69         M6P            West Toronto   
70         M9P               Etobicoke   
71         M1R             Scarborough   
72         M2R              North York   
73         M4R         Central Toronto   
74         M5R         Central Toronto   
75         M6R            West Toronto   
76         M7R             Mississauga   
77         M9R               Etobicoke   
78         M1S             Scarborough   
79         M4S         Central Toronto   
80         M5S        Downtown Toronto   
81         M6S            West Toronto   
82         M1T             Scarborough   
83         M4T         Central Toronto   
84         M5T        Downtown Toronto   
85         M1V             Scarborough   
86         M4V         Central Toronto   
87         M5V        Downtown Toronto   
88         M8V               Etobicoke   
89         M9V               Etobicoke   
90         M1W             Scarborough   
91         M4W        Downtown Toronto   
92         M5W  Downtown Toronto Stn A   
93         M8W               Etobicoke   
94         M9W

In [5]:
df.shape

(103, 3)

### Part 2 - Updating the data frame to include geographical coordinates 

#### Get the coordinates of each postal code in a seperate data frame.

In [38]:
path='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv'
df_coord = pd.read_csv(path)

df_coord.columns = ["PostalCode", "Latitude", "Longitude"]
#df_coord

#### Merge the two data frames to create a new one that contains all the data

In [7]:
data = pd.merge(df, df_coord)

#### Resulting dataframe and its dimension (Part 2 result)

In [8]:
data

PostalCode                 Borough  \
0          M3A              North York   
1          M4A              North York   
2          M5A        Downtown Toronto   
3          M6A              North York   
4          M7A            Queen's Park   
5          M9A               Etobicoke   
6          M1B             Scarborough   
7          M3B              North York   
8          M4B               East York   
9          M5B        Downtown Toronto   
10         M6B              North York   
11         M9B               Etobicoke   
12         M1C             Scarborough   
13         M3C              North York   
14         M4C               East York   
15         M5C        Downtown Toronto   
16         M6C                    York   
17         M9C               Etobicoke   
18         M1E             Scarborough   
19         M4E            East Toronto   
20         M5E        Downtown Toronto   
21         M6E                    York   
22         M1G             Scarborough   
23         M4G               East York   
24         M5G        Downtown Toronto   
25         M6G        Downtown Toronto   
26         M1H             Scarborough   
27         M2H              North York   
28         M3H              North York   
29         M4H               East York   
30         M5H        Downtown Toronto   
31         M6H            West Toronto   
32         M1J             Scarborough   
33         M2J              North York   
34         M3J              North York   
35         M4J  East York/East Toronto   
36         M5J        Downtown Toronto   
37         M6J            West Toronto   
38         M1K             Scarborough   
39         M2K              North York   
40         M3K              North York   
41         M4K            East Toronto   
42         M5K        Downtown Toronto   
43         M6K            West Toronto   
44         M1L             Scarborough   
45         M2L              North York   
46         M3L              North York   
47         M4L            East Toronto   
48         M5L        Downtown Toronto   
49         M6L              North York   
50         M9L              North York   
51         M1M             Scarborough   
52         M2M              North York   
53         M3M              North York   
54         M4M            East Toronto   
55         M5M              North York   
56         M6M                    York   
57         M9M              North York   
58         M1N             Scarborough   
59         M2N              North York   
60         M3N              North York   
61         M4N         Central Toronto   
62         M5N         Central Toronto   
63         M6N                    York   
64         M9N                    York   
65         M1P             Scarborough   
66         M2P              North York   
67         M4P         Central Toronto   
68         M5P         Central Toronto   
69         M6P            West Toronto   
70         M9P               Etobicoke   
71         M1R             Scarborough   
72         M2R              North York   
73         M4R         Central Toronto   
74         M5R         Central Toronto   
75         M6R            West Toronto   
76         M7R             Mississauga   
77         M9R               Etobicoke   
78         M1S             Scarborough   
79         M4S         Central Toronto   
80         M5S        Downtown Toronto   
81         M6S            West Toronto   
82         M1T             Scarborough   
83         M4T         Central Toronto   
84         M5T        Downtown Toronto   
85         M1V             Scarborough   
86         M4V         Central Toronto   
87         M5V        Downtown Toronto   
88         M8V               Etobicoke   
89         M9V               Etobicoke   
90         M1W             Scarborough   
91         M4W        Downtown Toronto   
92         M5W  Downtown Toronto Stn A   
93         M8W               Etobicoke   
94         M9W

In [9]:
data.shape

(103, 5)

### Part 3 - Clustering the neighborhoods in Toronto

#### Get only Borough with the word 'Toronto' 

In [11]:
toronto_data = data.loc[data['Borough'].str.contains("Toronto", case =False)]
toronto_data.reset_index(inplace=True,drop=True)
toronto_data

PostalCode                 Borough  \
0         M5A        Downtown Toronto   
1         M5B        Downtown Toronto   
2         M5C        Downtown Toronto   
3         M4E            East Toronto   
4         M5E        Downtown Toronto   
5         M5G        Downtown Toronto   
6         M6G        Downtown Toronto   
7         M5H        Downtown Toronto   
8         M6H            West Toronto   
9         M4J  East York/East Toronto   
10        M5J        Downtown Toronto   
11        M6J            West Toronto   
12        M4K            East Toronto   
13        M5K        Downtown Toronto   
14        M6K            West Toronto   
15        M4L            East Toronto   
16        M5L        Downtown Toronto   
17        M4M            East Toronto   
18        M4N         Central Toronto   
19        M5N         Central Toronto   
20        M4P         Central Toronto   
21        M5P         Central Toronto   
22        M6P            West Toronto   
23        M4R         Central Toronto   
24        M5R         Central Toronto   
25        M6R            West Toronto   
26        M4S         Central Toronto   
27        M5S        Downtown Toronto   
28        M6S            West Toronto   
29        M4T         Central Toronto   
30        M5T        Downtown Toronto   
31        M4V         Central Toronto   
32        M5V        Downtown Toronto   
33        M4W        Downtown Toronto   
34        M5W  Downtown Toronto Stn A   
35        M4X        Downtown Toronto   
36        M5X        Downtown Toronto   
37        M4Y        Downtown Toronto   
38        M7Y   East Toronto Business   

                                         Neighborhood   Latitude  Longitude  
0                           Regent Park, Harbourfront  43.654260 -79.360636  
1                            Garden District, Ryerson  43.657162 -79.378937  
2                                      St. James Town  43.651494 -79.375418  
3                                         The Beaches  43.676357 -79.293031  
4                                         Berczy Park  43.644771 -79.373306  
5                                  Central Bay Street  43.657952 -79.387383  
6                                            Christie  43.669542 -79.422564  
7                            Richmond, Adelaide, King  43.650571 -79.384568  
8                        Dufferin, Dovercourt Village  43.669005 -79.442259  
9                                  The Danforth  East  43.685347 -79.338106  
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
11                           Little Portugal, Trinity  43.647927 -79.419750  
12                       The Danforth West, Riverdale  43.679557 -79.352188  
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
15                     India Bazaar, The Beaches West  43.668999 -79.315572  
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
17                                    Studio District  43.659526 -79.340923  
18                                      Lawrence Park  43.728020 -79.388790  
19                                           Roselawn  43.711695 -79.416936  
20                                   Davisville North  43.712751 -79.390197  
21                           Forest Hill North & West  43.696948 -79.411307  
22                      High Park, The Junction South  43.661608 -79.464763  
23                                 North Toronto West  43.715383 -79.405678  
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
25                             Parkdale, Roncesvalles  43.648960 -79.456325  
26                                         Davisville  43.704324 -79.388790  
27                     University of Toronto, Harbord  43.662696 -79.400049  
28                                 Runnymede, Swansea  43.651571 -79.484450  
29                    

#### Shape of the resulting neighborhoods

In [12]:
toronto_data.shape
# We now have 39 neighborhoods

(39, 5)

#### Get data from FourSquare API for each neighborhood

In [13]:
CLIENT_ID = 'UJXTAUXSHD0COLY1WIO05G2TUNNO2EPEK0UOVGIHBUDUVVSN' # your Foursquare ID
CLIENT_SECRET = '5S53L4AFGWV5L25WEJNRSRFBYTLGLLWIHFOOXKCGCQAX2LWA' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value
RADIUS = 500


In [14]:
# Function to get nearby venues for given latitude and longitude 
def getNearbyVenues(postalCode, boroughs, names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for code, borough, name, lat, lng in zip(postalCode, boroughs, names, latitudes, longitudes):
        print(code, borough, name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            RADIUS, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            code,
            borough,
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['PostalCode',
                'Borough',
                'Neighborhood', 
                'Neighborhood Latitude', 
                'Neighborhood Longitude', 
                'Venue', 
                'Venue Latitude', 
                'Venue Longitude', 
                'Venue Category']
    
    return(nearby_venues)

In [15]:
toronto_venues = getNearbyVenues(postalCode=toronto_data['PostalCode'],
                                 boroughs=toronto_data['Borough'],
                                 names=toronto_data['Neighborhood'],
                                 latitudes=toronto_data['Latitude'],
                                 longitudes=toronto_data['Longitude']
                                )

M5A Downtown Toronto Regent Park, Harbourfront
M5B Downtown Toronto Garden District, Ryerson
M5C Downtown Toronto St. James Town
M4E East Toronto The Beaches
M5E Downtown Toronto Berczy Park
M5G Downtown Toronto Central Bay Street
M6G Downtown Toronto Christie
M5H Downtown Toronto Richmond, Adelaide, King
M6H West Toronto Dufferin, Dovercourt Village
M4J East York/East Toronto The Danforth  East
M5J Downtown Toronto Harbourfront East, Union Station, Toronto Islands
M6J West Toronto Little Portugal, Trinity
M4K East Toronto The Danforth West, Riverdale
M5K Downtown Toronto Toronto Dominion Centre, Design Exchange
M6K West Toronto Brockton, Parkdale Village, Exhibition Place
M4L East Toronto India Bazaar, The Beaches West
M5L Downtown Toronto Commerce Court, Victoria Hotel
M4M East Toronto Studio District
M4N Central Toronto Lawrence Park
M5N Central Toronto Roselawn
M4P Central Toronto Davisville North
M5P Central Toronto Forest Hill North & West
M6P West Toronto High Park, The Junction

In [16]:
toronto_venues.head()

PostalCode           Borough               Neighborhood  \
0        M5A  Downtown Toronto  Regent Park, Harbourfront   
1        M5A  Downtown Toronto  Regent Park, Harbourfront   
2        M5A  Downtown Toronto  Regent Park, Harbourfront   
3        M5A  Downtown Toronto  Regent Park, Harbourfront   
4        M5A  Downtown Toronto  Regent Park, Harbourfront   

   Neighborhood Latitude  Neighborhood Longitude                   Venue  \
0               43.65426              -79.360636        Roselle Desserts   
1               43.65426              -79.360636           Tandem Coffee   
2               43.65426              -79.360636  Cooper Koo Family YMCA   
3               43.65426              -79.360636     Body Blitz Spa East   
4               43.65426              -79.360636          Impact Kitchen   

   Venue Latitude  Venue Longitude       Venue Category  
0       43.653447       -79.362017               Bakery  
1       43.653559       -79.361809          Coffee Shop  
2       43.653249       -79.358008  Distribution Center  
3       43.654735       -79.359874                  Spa  
4       43.656369       -79.356980           Restaurant

In [17]:
toronto_venues.shape

(1584, 9)

Check the venue categories 

In [18]:
toronto_venues.groupby('Venue Category').count()


PostalCode  Borough  Neighborhood  \
Venue Category                                                       
Airport                                   1        1             1   
Airport Food Court                        1        1             1   
Airport Lounge                            2        2             2   
Airport Service                           3        3             3   
Airport Terminal                          2        2             2   
American Restaurant                      15       15            15   
Antique Shop                              3        3             3   
Aquarium                                  5        5             5   
Art Gallery                              13       13            13   
Art Museum                                2        2             2   
Arts & Crafts Store                       4        4             4   
Asian Restaurant                         10       10            10   
Auto Workshop                             1        1             1   
BBQ Joint                                 2        2             2   
Baby Store                                1        1             1   
Bagel Shop                                2        2             2   
Bakery                                   39       39            39   
Bank                                     13       13            13   
Bar                                      26       26            26   
Baseball Stadium                          2        2             2   
Basketball Stadium                        4        4             4   
Beach                                     1        1             1   
Bed & Breakfast                           1        1             1   
Beer Bar                                 18       18            18   
Beer Store                                4        4             4   
Belgian Restaurant                        2        2             2   
Bistro                                    4        4             4   
Board Shop                                1        1             1   
Boat or Ferry                             1        1             1   
Bookstore                                16       16            16   
Brazilian Restaurant                      2        2             2   
Breakfast Spot                           16       16            16   
Brewery                                  12       12            12   
Bubble Tea Shop                           9        9             9   
Building                                  1        1             1   
Burger Joint                             14       14            14   
Burrito Place                            10       10            10   
Bus Line                                  1        1             1   
Business Service                          1        1             1   
Butcher                                   2        2             2   
Café                                     85       85            85   
Cajun / Creole Restaurant                 1        1             1   
Candy Store                               1        1             1   
Caribbean Restaurant                      5        5             5   
Cheese Shop                               6        6             6   
Chinese Restaurant                        6        6             6   
Chocolate Shop                            1        1             1   
Church                                    1        1             1   
Climbing Gym                              1        1             1   
Clothing Store                           18       18            18   
Cocktail Bar                             13       13            13   
Coffee Shop                             146      146           146   
College Arts Building                     1        1             1   
College Gym                               1        1             1   
College Rec Center                        1        1             1   
Colombian Restaurant                      2

Rename the Venue category called 'Neighborhood' to 'Neighborhood Area' to avoid ambiguity

In [19]:
toronto_venues.loc[toronto_venues['Venue Category'].str.contains("Neighborhood", case =False), 'Venue Category'] = "Neighborhood Area"

Check that they have been changed 

In [20]:
toronto_venues.loc[toronto_venues['Venue Category'].str.contains("Neighborhood Area", case =False), 'Venue Category'] 

232    Neighborhood Area
379    Neighborhood Area
487    Neighborhood Area
921    Neighborhood Area
Name: Venue Category, dtype: object

In [21]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 228 uniques categories.


In [22]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()


Neighborhood  Airport  Airport Food Court  Airport Lounge  \
0  Regent Park, Harbourfront        0                   0               0   
1  Regent Park, Harbourfront        0                   0               0   
2  Regent Park, Harbourfront        0                   0               0   
3  Regent Park, Harbourfront        0                   0               0   
4  Regent Park, Harbourfront        0                   0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0            0           0                    0                 0   
1         0            0           0                    0                 0   
2         0            0           0                    0                 0   
3         0            0           0                    0                 0   
4         0            0           0                    0                 0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       1     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bistro  Board Shop  Boat or Ferry  \
0           0                   0       0           0              0   
1           0                   0       0           0              0   
2           0                   0       0           0              0   
3           0                   0       0           0              0   
4           0                   0       0           0              0   

   Bookstore  Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  \
0          0                     0               0        0                0   
1          0                     0               0        0                0   
2          0                     0               0        0                0   
3          0                     0               0        0                0   
4          0                     0               0        0                0   

   Building  Burger Joint  Burrito Place  Bus Line  Business Service  Butcher  \
0         0             0              0         0                 0        0   
1         0             0              0         0                 0        0   
2         0             0              0         0                 0        0   
3         0             0              0         0                 0        0   
4         0             0              0         0                 0        0   

   Café  Cajun / Creole Restaurant  Candy Store  Caribbean Restaurant  \
0     0                          0            0                     0   
1     0                          0            0                     0   
2     0                          0            0                     0   
3     0                   

In [23]:
toronto_onehot.shape

(1584, 229)

Now there are 229 columns (228 categories + 1 column for neighbourhood). Next ereset the index and confirm the shape

In [39]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
#toronto_grouped

In [25]:
toronto_grouped.shape
# This confirms that we still have 39 neighborhoods

(39, 229)

This would print each neighborhood with the top 5 most common venue. The print statements are commented for ease of reading the notebook

In [43]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    #print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    #print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    #print('\n')

Function to sort venues in descending order 

In [27]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Create the new dataframe and display the top 10 venues for each neighborhood.


In [28]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
3                                 Central Bay Street           Coffee Shop   
4                                           Christie         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery          Cocktail Bar              Pharmacy   
1           Coffee Shop        Breakfast Spot             Pet Store   
2        Airport Lounge      Airport Terminal       Harbor / Marina   
3        Sandwich Place                  Café    Italian Restaurant   
4                  Café                  Park            Baby Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Farmers Market    Seafood Restaurant    Italian Restaurant   
1               Stadium         Burrito Place            Restaurant   
2      Sculpture Garden    Airport Food Court                   Bar   
3            Restaurant           Salad Place   Japanese Restaurant   
4           Coffee Shop             Nightclub            Restaurant   

  8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0            Restaurant               Beer Bar            Cheese Shop  
1          Climbing Gym  Performing Arts Venue                 Bakery  
2         Boat or Ferry    Rental Car Location                Airport  
3          Burger Joint        Bubble Tea Shop       Ramen Restaurant  
4    Italian Restaurant            Candy Store                Dog Run

Run _k_-means to cluster the neighborhood into 5 clusters.

In [29]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 1, 2, 0, 0, 0, 0, 0, 0, 3, 4, 0, 0, 0], dtype=int32)

Create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.


In [30]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode           Borough               Neighborhood   Latitude  \
0        M5A  Downtown Toronto  Regent Park, Harbourfront  43.654260   
1        M5B  Downtown Toronto   Garden District, Ryerson  43.657162   
2        M5C  Downtown Toronto             St. James Town  43.651494   
3        M4E      East Toronto                The Beaches  43.676357   
4        M5E  Downtown Toronto                Berczy Park  43.644771   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.360636               0           Coffee Shop                  Café   
1 -79.378937               0           Coffee Shop        Clothing Store   
2 -79.375418               0           Coffee Shop                  Café   
3 -79.293031               3                 Trail     Neighborhood Area   
4 -79.373306               0           Coffee Shop                Bakery   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                Bakery                   Pub                  Park   
1        Cosmetics Shop   Japanese Restaurant       Bubble Tea Shop   
2          Cocktail Bar            Restaurant              Beer Bar   
3     Health Food Store                   Pub   Dumpling Restaurant   
4          Cocktail Bar              Pharmacy        Farmers Market   

  6th Most Common Venue      7th Most Common Venue 8th Most Common Venue  \
0               Theater             Breakfast Spot            Restaurant   
1    Italian Restaurant  Middle Eastern Restaurant                  Café   
2        Cosmetics Shop                      Diner        Lingerie Store   
3   Distribution Center                    Dog Run      Doner Restaurant   
4    Seafood Restaurant         Italian Restaurant            Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0        Cosmetics Shop             Shoe Store  
1                 Hotel      Electronics Store  
2                Bakery    Moroccan Restaurant  
3            Donut Shop            Yoga Studio  
4              Beer Bar            Cheese Shop

#### Use geopy to get the longitude and latitude of Toronto

In [31]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="toronto_explore")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [44]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.8).add_to(map_clusters)
       
map_clusters

#### Examining each cluster

#### Cluster 1

In [33]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[0,1] + list(range(5, toronto_merged.shape[1]))]]

PostalCode                 Borough  Cluster Labels 1st Most Common Venue  \
0         M5A        Downtown Toronto               0           Coffee Shop   
1         M5B        Downtown Toronto               0           Coffee Shop   
2         M5C        Downtown Toronto               0           Coffee Shop   
4         M5E        Downtown Toronto               0           Coffee Shop   
5         M5G        Downtown Toronto               0           Coffee Shop   
6         M6G        Downtown Toronto               0         Grocery Store   
7         M5H        Downtown Toronto               0           Coffee Shop   
8         M6H            West Toronto               0              Pharmacy   
10        M5J        Downtown Toronto               0           Coffee Shop   
11        M6J            West Toronto               0                   Bar   
12        M4K            East Toronto               0      Greek Restaurant   
13        M5K        Downtown Toronto               0           Coffee Shop   
14        M6K            West Toronto               0                  Café   
15        M4L            East Toronto               0                  Park   
16        M5L        Downtown Toronto               0           Coffee Shop   
17        M4M            East Toronto               0           Coffee Shop   
18        M4N         Central Toronto               0                  Park   
20        M4P         Central Toronto               0                 Hotel   
22        M6P            West Toronto               0                  Café   
23        M4R         Central Toronto               0           Coffee Shop   
24        M5R         Central Toronto               0           Coffee Shop   
25        M6R            West Toronto               0        Breakfast Spot   
26        M4S         Central Toronto               0        Sandwich Place   
27        M5S        Downtown Toronto               0                  Café   
28        M6S            West Toronto               0                  Café   
30        M5T        Downtown Toronto               0                   Bar   
31        M4V         Central Toronto               0           Coffee Shop   
32        M5V        Downtown Toronto               0       Airport Service   
34        M5W  Downtown Toronto Stn A               0           Coffee Shop   
35        M4X        Downtown Toronto               0           Coffee Shop   
36        M5X        Downtown Toronto               0           Coffee Shop   
37        M4Y        Downtown Toronto               0           Coffee Shop   
38        M7Y   East Toronto Business               0                  Park   

   2nd Most Common Venue 3rd Most Common Venue          4th Most Common Venue  \
0                   Café                Bakery                            Pub   
1         Clothing Store        Cosmetics Shop            Japanese Restaurant   
2                   Café          Cocktail Bar                     Restaurant   
4                 Bakery          Cocktail Bar                       Pharmacy   
5         Sandwich Place                  Café             Italian Restaurant   
6                   Café                  Park                     Baby Store   
7                   Café            Restaurant                Thai Restaurant   
8                 Bakery           Music Venue                           Café   
10              Aquarium            Restaurant                           Café   
11      Asian Restaurant                  Café                     Restaurant   
12           Coffee Shop    Italian Restaurant         Furniture / Home Store   
13                  Café                 Hotel                     Restaurant   
14           Coffee Shop        Breakfast Spot                      Pet Store   
15  Fast Food Restaurant        Ice Cream Shop                     Board Shop   
16            Restaurant                  Café                          Hotel   
17               Bre

#### Cluster 2

In [34]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[0,1] + list(range(5, toronto_merged.shape[1]))]]

PostalCode           Borough  Cluster Labels 1st Most Common Venue  \
21        M5P   Central Toronto               1                  Park   
29        M4T   Central Toronto               1          Tennis Court   
33        M4W  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21                 Trail         Jewelry Store      Sushi Restaurant   
29            Playground                  Park                 Diner   
33            Playground                 Trail           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
21           Yoga Studio   Distribution Center    Falafel Restaurant   
29           Event Space  Ethiopian Restaurant           Escape Room   
33                 Diner           Event Space  Ethiopian Restaurant   

   8th Most Common Venue        9th Most Common Venue  \
21           Event Space         Ethiopian Restaurant   
29     Electronics Store  Eastern European Restaurant   
33           Escape Room            Electronics Store   

         10th Most Common Venue  
21                  Escape Room  
29          Dumpling Restaurant  
33  Eastern European Restaurant

#### Cluster 3

In [35]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[0,1] + list(range(5, toronto_merged.shape[1]))]]

PostalCode          Borough  Cluster Labels 1st Most Common Venue  \
19        M5N  Central Toronto               2  Fast Food Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19                Garden        Farmers Market    Falafel Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
19           Event Space  Ethiopian Restaurant           Escape Room   

   8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
19     Electronics Store  Eastern European Restaurant    Dumpling Restaurant

#### Cluster 4

In [36]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[0,1] + list(range(5, toronto_merged.shape[1]))]]

PostalCode       Borough  Cluster Labels 1st Most Common Venue  \
3        M4E  East Toronto               3                 Trail   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
3     Neighborhood Area     Health Food Store                   Pub   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
3   Dumpling Restaurant   Distribution Center               Dog Run   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
3      Doner Restaurant            Donut Shop            Yoga Studio

#### Cluster 5

In [37]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[0,1] + list(range(5, toronto_merged.shape[1]))]]

PostalCode                 Borough  Cluster Labels 1st Most Common Venue  \
9        M4J  East York/East Toronto               4          Intersection   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
9     Convenience Store                  Park           Yoga Studio   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
9        Farmers Market    Falafel Restaurant           Event Space   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
9  Ethiopian Restaurant           Escape Room      Electronics Store